src/embed_ygm.cpp

```diff
@@ -0,0 +1,162 @@
+
+  double total_update_time{0.0};
+
+  auto begin_stats = world.stats_snapshot();
+
+  for (int trial = 0; trial < params.num_trials; ++trial) {
+    std::vector<std::pair<uint64_t, uint64_t>> edges(params.local_edge_count);
+
+    for (int64_t i = 0; i < params.local_edge_count; ++i) {
+      edges[i] = {dist(gen), dist(gen)};
+    }
+
+    world.barrier();
+
+    world.set_track_stats(true);
+
+    ygm::timer update_timer{};
+
+    for (const auto &edge : edges) {
+      dsk.async_update(edge.first, edge.second);
+      // dsk.async_insert(edge.second, edge.first);
+    }
+    dsk.compactify();
+
+    world.barrier();
+
+    world.set_track_stats(false);
+
+    double trial_time = update_timer.elapsed();
+
+    if (world.rank0()) {
+      double trial_rate = params.local_edge_count * world.size() / trial_time /
+                          (1000 * 1000 * 1000);
+
+      std::cout << ", " << trial_time << ", " << trial_rate;
+    }
+
+    total_update_time += trial_time;
+  }
+
+  uint64_t global_bytes = world.global_bytes_sent();
+  uint64_t global_message_bytes =
+      params.local_edge_count * world.size() * params.num_trials * 8;
+
+  if (world.rank0()) {
+    double average_rate = params.local_edge_count * world.size() *
+                          params.num_trials / total_update_time /
+                          (1000 * 1000 * 1000);
+
+    std::cout << std::endl;
+  }
+
+  auto experiment_stats = world.stats_snapshot() - begin_stats;
+  write_stats_files(world, experiment_stats, params.stats_output_prefix);
+}
+
+int main(int argc, char **argv) {
+  int provided;
+  MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
+  if (provided != MPI_THREAD_MULTIPLE) {
+    throw std::runtime_error(
+        "MPI_Init_thread: MPI_THREAD_MULTIPLE not provided.");
+  }
+
+  {
+    int ygm_buffer_capacity = atoi(argv[1]);
+
+    ygm::comm world(MPI_COMM_WORLD, ygm_buffer_capacity);
+
+    int         range_size(atoi(argv[2]));
+    int         log_vertex_count{atoi(argv[3])};
+    int         log_local_edge_count{atoi(argv[4])};
+    size_t      compaction_threshold(atoll(argv[5]));
+    size_t      promotion_threshold(atoll(argv[6]));
+    int         num_trials{atoi(argv[7])};
+    uint32_t    seed(atoi(argv[8]));
+    std::string stats_output_prefix(argv[9]);
+
+    size_t vertex_count     = ((size_t)1) << log_vertex_count;
+    size_t local_edge_count = ((size_t)1) << log_local_edge_count;
+    // world.cout0("Creating global table with size ", global_table_size);
+    if (world.rank0()) {
+      std::cout << world.layout().node_size() << ", "
+                << world.layout().local_size() << ", " << ygm_buffer_capacity
+                << ", " << world.routing_protocol() << ", " << range_size
+                << ", " << vertex_count << ", "
+                << local_edge_count * world.size() << ", "
+                << compaction_threshold << ", " << promotion_threshold;
```

Comment:
Can you print the seed after the promotion threshold to match the headers?

Reply:
Oops! fixed.